adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1320 @@
+	} else {
+		return DefaultCountryName
+	}
+}
+
+// get device id  include oaid gaid imei.
+// For example: prebid mobile -> TargetingParams.addUserData("imei", "imei-test");
+func getDeviceID(device *device, openRTBRequest *openrtb2.BidRequest) (err error) {
+	if openRTBRequest.User == nil {
+		return errors.New("getDeviceID: openRTBRequest.User is nil.")
+	}
+	if openRTBRequest.User.Ext == nil {
+		return errors.New("getDeviceID: openRTBRequest.User.Ext is nil.")
+	}
+	var extUserDataHuaweiAds openrtb_ext.ExtUserDataHuaweiAds
+	if err := json.Unmarshal(openRTBRequest.User.Ext, &extUserDataHuaweiAds); err != nil {
+		return errors.New("Unmarshal: openRTBRequest.User.Ext -> extUserDataHuaweiAds failed")
+	}
+	var deviceId = extUserDataHuaweiAds.Data
+	if len(deviceId.Imei) == 0 && len(deviceId.Gaid) == 0 && len(deviceId.Oaid) == 0 {
+		return errors.New("getDeviceID: Imei ,Oaid, Gaid are all empty.")
+	}
+	if len(deviceId.Oaid) > 0 {
+		device.Oaid = deviceId.Oaid[0]
+	}
+	if len(deviceId.Gaid) > 0 {
+		device.Gaid = deviceId.Gaid[0]
+	}
+	if len(deviceId.Imei) > 0 {
+		device.Imei = deviceId.Imei[0]
+	}
+	if len(deviceId.ClientTime) > 0 {
+		device.ClientTime = getClientTime(deviceId.ClientTime[0])
+	}
+	// IsTrackingEnabled = 1 - DNT
+	if openRTBRequest.Device != nil && openRTBRequest.Device.DNT != nil {
+		if device.Oaid != "" {
+			device.IsTrackingEnabled = strconv.Itoa(1 - int(*openRTBRequest.Device.DNT))
+		}
+		if device.Gaid != "" {
+			device.GaidTrackingEnabled = strconv.Itoa(1 - int(*openRTBRequest.Device.DNT))
+		}
+	}
+	return nil
+}
+
+// get network information for HuaweiAds request, include Carrier, Mcc, Mnc
+func getHuaweiAdsReqNetWorkInfo(request *huaweiAdsRequest, openRTBRequest *openrtb2.BidRequest) {
+	if openRTBRequest.Device != nil {
+		var network network
+		if openRTBRequest.Device.ConnectionType != nil {
+			network.Type = int32(*openRTBRequest.Device.ConnectionType)
+		} else {
+			network.Type = DefaultUnknownNetworkType
+		}
+
+		var cellInfos []cellInfo
+		if openRTBRequest.Device.MCCMNC != "" {
+			var arr = strings.Split(openRTBRequest.Device.MCCMNC, "-")
+			network.Carrier = 0
+			if len(arr) >= 2 {
+				cellInfos = append(cellInfos, cellInfo{
+					Mcc: arr[0],
+					Mnc: arr[1],
+				})
+				var str = arr[0] + arr[1]
+				if str == "46000" || str == "46002" || str == "46007" {
+					network.Carrier = 2
+				} else if str == "46001" || str == "46006" {
+					network.Carrier = 1
+				} else if str == "46003" || str == "46005" || str == "46011" {
+					network.Carrier = 3
+				} else {
+					network.Carrier = 99
+				}
+			}
+		}
+		network.CellInfo = cellInfos
+		request.Network = network
+	}
+}
+
+// get regs information for HuaweiAds request
+func getHuaweiAdsReqRegsInfo(request *huaweiAdsRequest, openRTBRequest *openrtb2.BidRequest) {
+	if openRTBRequest.Regs != nil && openRTBRequest.Regs.COPPA >= 0 {
+		var regs regs
+		regs.Coppa = int32(openRTBRequest.Regs.COPPA)
+		request.Regs = regs
```

Comment:
Please add test coverage for when coppa is specified.

Reply:
added in banner1.json